ddprofiler/app/src/main/java/ddprofiler/preanalysis/PreAnalyzer.java

```diff
@@ -173,84 +175,75 @@ private void calculateDataTypes(Map<Attribute, List<String>> data) {
-            }
-        }
+                        // Check whether the value match any of the temporal patterns.
+                        // For temporal, we allow early break because each pattern is specific enough.
+                        for (Map.Entry<String, Pattern[]> patternsEntry : TEMPORAL_PATTERNS.entrySet()) {
+                            for (Pattern temporalPattern : patternsEntry.getValue()) {
+                                if (temporalPattern.matcher(cleanedValue).matches()) {
+                                    String granularity = patternsEntry.getKey();
+                                    attribute.setColumnSemanticType(AttributeSemanticType.TEMPORAL);
+                                    attribute.getColumnSemanticTypeDetails().put("granularity", granularity);
+                                    break valueLoop;
+                                }
+                            }
+                        }
 
-        return granularityMatchCounts.entrySet().stream()
-                .max(Entry.comparingByValue()).map(Entry::getKey).orElse("unknown");
-    }
+                        // Check whether the value match any of the spatial patterns.
+                        // For spatial, we consider partial matching with a threshold.
+                        for (Map.Entry<String, Pattern[]> patternsEntry : SPATIAL_PATTERNS.entrySet()) {
+                            for (Pattern spatialPattern : patternsEntry.getValue()) {
+                                if (spatialPattern.matcher(cleanedValue).find()) {
+                                    String granularity = patternsEntry.getKey();
+                                    if (spatialMatchCount.containsKey(granularity)) {
+                                        spatialMatchCount.put(
+                                                granularity,
+                                                spatialMatchCount.get(granularity) + 1);
+                                    } else {
+                                        spatialMatchCount.put(granularity, 1L);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
 
-    private String checkTemporalGranularity(String value) {
-        for (Map.Entry<String, Pattern[]> entry : TEMPORAL_PATTERNS.entrySet()) {
-            for (Pattern pattern : entry.getValue()) {
-                if (pattern.matcher(value).matches()) {
-                    return entry.getKey();
+                // Check whether any spatial pattern matched > 50% of the values to determine the types.
+                // Assumption: no spatial and temporal patterns have >50% matches at the same time.
+                for (String spatialGranularity : spatialMatchCount.keySet()) {
+                    if ((long) spatialMatchCount.get(spatialGranularity) / validValuesCount > 0.5) {
```

Comment:
```suggestion
                    if ((double) spatialMatchCount.get(spatialGranularity) / validValuesCount > 0.5) {
```

Reply:
This is an integer division in floating-point context. The result of this equation will always be integer. The variable validValuesCount doesn't need to be long since the readRow function will only read 1000 rows at a time anyway.

Also, is there a chance that there will be a division by zero if there is no valid values?

Reply:
> The result of this equation will always be integer

Thank you for pointing out a bug! It improves the performance:

- SPATIAL Label's Precision: 0.9861111111111112
- TEMPORAL Label's Precision: 0.9707207207207207
---------------------------------------------------
- SPATIAL Label's Recall: 0.9906976744186047
- TEMPORAL Label's Recall: 0.9685393258426966

> since the readRow function will only read 1000 rows at a time anyway.

It can change depending on the configuration of ProfilerConfig (`NUM_RECORD_READ`), although it might be an unlikely configuration for it to be greater than the max of int. Should I change it to int? @snowgy

> is there a chance that there will be a division by zero if there is no valid values?

Yes, I just realized that it is possible. I will modify the code to check for that.